### PR TITLE
more resilient subsystems: av-store

### PR DESCRIPTION
**Pull request by ordian**
Part of #1602.

I've decided to split the work for each subsystem first to validate the approach and second to parallelize the efforts.

It seems there are at least 3 classes of errors we should handle more gracefully:
1. RuntimeAPI errors (we can remove `impl std::error:Error for RuntimeApiError` later to ensure that)
2. Send errors, e.g. ~~channel capacity errors~~ closed channel.
3. Requests to job-based subsystems can fail due to race conditions mentioned in https://github.com/paritytech/polkadot/pull/1583#issuecomment-675028018.

Errors like `ChainAPIError` or something that breaks internal logic should probably be hard errors.

Please let me know if I'm missing something (I certainly am).

### CODE REVIEW

**Comment by bkchr** [comment]
>     2\. Send errors, e.g. channel capacity errors

Don't we mainly use `send().await` and so we should not get any capacity errors?

**Comment by ordian** [comment]
> > ```
> > 2\. Send errors, e.g. channel capacity errors
> > ```
> 
> Don't we mainly use `send().await` and so we should not get any capacity errors?

Overseer handler has a bounded channel ([1024](https://github.com/paritytech/polkadot/blob/60b9e368e68ccd0cd149173f6811e88ee7f4ebab/node/overseer/src/lib.rs#L98), so e.g.
https://github.com/paritytech/polkadot/blob/60b9e368e68ccd0cd149173f6811e88ee7f4ebab/node/core/av-store/src/lib.rs#L640
can fail due to capacity error, but there are other cases as well.

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
> > > ```
> > > 2\. Send errors, e.g. channel capacity errors
> > > ```
> > 
> > 
> > Don't we mainly use `send().await` and so we should not get any capacity errors?
> 
> Overseer handler has a bounded channel ([1024](https://github.com/paritytech/polkadot/blob/60b9e368e68ccd0cd149173f6811e88ee7f4ebab/node/overseer/src/lib.rs#L98), so e.g.
> 
> https://github.com/paritytech/polkadot/blob/60b9e368e68ccd0cd149173f6811e88ee7f4ebab/node/core/av-store/src/lib.rs#L640
> 
> can fail due to capacity error, but there are other cases as well.

https://docs.rs/futures/0.3.7/futures/sink/trait.SinkExt.html#method.send

Or do I miss something? The whole idea why this function is async is that you wait until there is capacity to send data again (back pressure).

**Comment by ordian** [comment]
There is `SendError::Full` https://docs.rs/futures/0.3.7/futures/channel/mpsc/struct.SendError.html, so I assume that's not the case.

**Comment by coriolinus** [comment]
I imagine that that type of `SendError` is used when using [non-async sending methods](https://docs.rs/futures/0.3.7/futures/channel/mpsc/struct.Sender.html#method.start_send).

**Comment by bkchr** [comment]
Yeah that is right, but the whole point of this function being async is that it waits for capacity. Check here:
https://docs.rs/futures-util/0.3.7/src/futures_util/sink/send.rs.html#36

They first call `poll_ready` if that returns Ready, they know that they can push a message. (Remember that each sender get a fixed slot)

File: node/core/av-store/src/lib.rs
Comment (pepyakin) [inline node/core/av-store/src/lib.rs:66]: Nit: It's `ChainApi`, [accronyms count as one word](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html).

File: node/core/av-store/src/lib.rs
Comment (pepyakin) [inline node/core/av-store/src/lib.rs:716]: can we make these distinct? I guess, it may be useful to know in which branch this log was emitted exactly.

File: node/core/av-store/src/lib.rs
Comment (pepyakin) [inline node/core/av-store/src/lib.rs:73]: If I had to handle this error I'd be curious what I can or cannot do with it. Can we mention that these must not lead to bringing down the whole subsystem?

**Comment by ordian** [comment]
> Yeah that is right, but the whole point of this function being async is that it waits for capacity. Check here:
> https://docs.rs/futures-util/0.3.7/src/futures_util/sink/send.rs.html#36
> 
> They first call `poll_ready` if that returns Ready, they know that they can push a message. (Remember that each sender get a fixed slot)

I stand corrected, thanks. Was confused by `tx.send(()).await` returning `Result<(), SendError>`. 
Still we want to account for channel send errors if the channel is closed.

File: node/core/av-store/src/lib.rs
Comment (drahnr) [reply node/core/av-store/src/lib.rs]: I'd suggest either adding a `tag: &'static str` to the error variant or introducing more error variants.

**Comment by bkchr** [comment]
But if the overseer channel is closed it means the node is shutting down? Or not?

**Comment by ordian** [comment]
> But if the overseer channel is closed it means the node is shutting down? Or not?

yes, well, if a subsystem cannot communicate with the overseer how does it even return the error :P 

but once the overseer relayed a message, there can be a channel between two subsystems

File: node/core/av-store/src/lib.rs
Comment (drahnr) [inline node/core/av-store/src/lib.rs:75]: Is this subsystem specific? Or is this generic over all subsystems? I think it is.

File: node/core/av-store/src/lib.rs
Comment (drahnr) [inline node/core/av-store/src/lib.rs:605]: This would be a common thing too I assume? Could be a generic `fn log_and_drop<E: ..., R: Try<..>>(e: E) -> R`

File: node/core/av-store/src/lib.rs
Comment (drahnr) [inline node/core/av-store/src/lib.rs:650]: If we know that all that is returned here, we should probably let the caller decide on what to do with the error rather than handling most of them within this `fn` already?

Comment (ordian) [reply node/core/av-store/src/lib.rs:650]: The caller is `process_block_activated` in this subsystem, it's handling the error. Not sure what exactly you're suggesting?

Comment (drahnr) [reply node/core/av-store/src/lib.rs:650]: In the following lines quite a few channel send error cases are silently dropped. If we know that all this fn returns is `NonFatalError`s anyways, we can also return those and let the caller handle the printing. It's a nit :) - generally I just prefer to have nicely testable `fn`s with well defined bounds (here it gets fuzzy with what non-fatal error is logged and which is returned).

Comment (ordian) [reply node/core/av-store/src/lib.rs:650]: In case of `process_message`, it actually returns hard errors (db or `get_chunk` failures), but silences `NonFatalError`s.
Are you suggesting we return something like `Result<Result<_, Error>, NonFatalError>` instead? It would lose the context that we have in `process_message` when logging errors.

Comment (coriolinus) [reply node/core/av-store/src/lib.rs:650]: An alternative would be something like

```rust
trait PolkadotError {
  fn is_fatal(&self) -> bool;
}
```

which, combined with requiring subsystems to return a `PolkadotError`, would mean that we could get away with just a single error type per subsystem, and let the overseer log and ignore all non-fatal errors.

Comment (drahnr) [reply node/core/av-store/src/lib.rs:650]: Downside would be to either require a macro to `#[soft]` (or so) to some error variants or impl the trait manually all over the place (juck).

Comment (coriolinus) [reply node/core/av-store/src/lib.rs:650]: I don't think that a macro which lets us tag certain variants as `#[soft]` would be that bad to use in practice, and it would be good fun to write the macro.

The main problem with this idea IMO is that all errors are hard in constructors, even those tagged as soft, but I don't see that as being a huge deal.

Comment (ordian) [reply node/core/av-store/src/lib.rs:650]: Every subsystem has to differentiate between the fatal and non-fatal error in order to decide when to shutdown in `fn run`, so they might as well log the non-fatal errors. 

I think it would be nice to utilize the type system somehow to ensure we can't accidentally shutdown on `NonFatalError` error, e.g. by not implementing `std::error:Error` or something.

**Comment by ordian** [comment]
After discussion in Element with the parachain team, we decided to make all subsystem errors non-fatal and never shutdown a subsystem. I'll address that for `av-store` subsystem in this PR.

**Review by bkchr — CHANGES_REQUESTED** [review]
Please add a test that ensures that the subsystem doesn't stop on an invalid runtime api request.

File: node/core/av-store/src/lib.rs
Comment (ordian) [inline node/core/av-store/src/lib.rs:66]: wait, did we merge these changes as part of #1864? :sweat_smile:

**Comment by ordian** [comment]
> Please add a test that ensures that the subsystem doesn't stop on an invalid runtime api request.

with the split of `run` into `run` and `run_iteration`, this is no longer required, but added in a599805 anyway

File: node/core/av-store/src/lib.rs
Comment (bkchr) [reply node/core/av-store/src/lib.rs:66]: Yeah :( I thought we would wait until this is merged, but I required some up-to-date code to see where it breaks.

File: node/core/av-store/src/lib.rs
Comment (ordian) [inline node/core/av-store/src/lib.rs:476]: this function can be extracted (as a generic) into utils module

File: node/core/av-store/src/lib.rs
Comment (rphmeier) [inline node/core/av-store/src/lib.rs:495]: It doesn't apply to this subsystem, but for other subsystems we need to be careful that `run_iteration` has the entire state of the subsystem passed into it. The more functional subsystems might require a few parameters or a refactoring to a larger `State` struct.

File: node/core/av-store/src/lib.rs
Comment (rphmeier) [reply node/core/av-store/src/lib.rs:476]: let's wait until we're aware of the exact patterns we see

**Review by rphmeier — APPROVED** [review]

**Pull request merged**
